### PR TITLE
arch/risc-v/bl602 : add pwm onshot watchdog driver.

### PR DESCRIPTION
**Pull request by liangzhanggb**
## Summary
arch/risc-v/bl602 : add pwm onshot watchdog driver

## Impact
N/A

## Testing
config:
1. → System Type → BL602 Peripheral Support → TIMER1
2. → Device Drivers → Timer Driver Support → PWM Driver Support → (5)Number of Output Channels Per Timer
3. → Device Drivers → Timer Driver Support → Oneshot timer drive
4. → Device Drivers → Timer Driver Support → Watchdog Timer Support
5. → RTOS Features → Performance Monitoring → Enable CPU load monitoring → Use external clock
6. → Application Configuration → Examples → Watchdog Timer example
7. → Application Configuration → Examples → Pulse width modulation (PWM) example 
result:
out

### CODE REVIEW

File: arch/risc-v/src/bl602/Kconfig [hunk 1]
```diff
@@ -34,4 +34,35 @@ config BL602_TIMER0
...
+
+config BL602_PWM0
+	bool "PWM0"
+	default y
+    depends on PWM 
+    select PWM_MULTICHAN
+
 endmenu
+
+config BL602_SPIFLASH
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/Kconfig:46]: @liangzhanggb Please split the SPI related change to another PR.

Comment (liangzhanggb) [reply arch/risc-v/src/bl602/Kconfig]: Thank you for your reminder, I will delete it immediately

**Comment by btashton** [comment]
@liangzhanggb  @xiaoxiang781216  I will review and test this evening. First glace there are some things that need to be moved.  The hardware/<ip>.h file should only have register information. I see some structs defined there that should be moved to the driver header in the next level up.

**Comment by liangzhanggb** [comment]
> @liangzhanggb @xiaoxiang781216 I will review and test this evening. First glace there are some things that need to be moved. The hardware/.h file should only have register information. I see some structs defined there that should be moved to the driver header in the next level up.

I have fixed it, Thank you very much

**Review by btashton — CHANGES_REQUESTED** [review]
@liangzhanggb I commented on some issues, but otherwise this looks very good.

I tested  (after a couple fixes I called out) the wdt and pwm with two channels enabled and the timing looks good.

```
nsh> pwm -f 20000 -c 1 -c 2 -d 20 -d 80
nsh: pwm: too many arguments
pwm_main: starting output with frequency: 20000 channel: 1 duty: 00003333 channel: 2 duty: 0000cccc
pwm_main: stopping output
nsh> 
```
image

I would also recommend that you add a new config for the board to highlight the timer features.  This is the one I generated if you want to use it.  This also provides a little more code coverage of the drivers to the CI system.

[defconfig.txt](https://github.com/apache/incubator-nuttx/files/5774565/defconfig.txt)

File: arch/risc-v/src/bl602/bl602_start.c [hunk 2]
```diff
@@ -97,9 +86,7 @@ uint32_t boot2_get_flash_addr(void)
 {
```
[the remark this answers is not in the document]
Comment (liangzhanggb) [reply arch/risc-v/src/bl602/bl602_start.c:86]: I have fixed it, Thank you very much

File: arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+  bl602_configgpio(BOARD_PWM_CH2_PIN);
+  pwm_init(2, BL_PWM_FREQ_DEFAULT);
+#endif
+
+#ifdef BOARD_PWM_CH3_PIN
+  bl602_configgpio(BOARD_PWM_CH3_PIN);
+  pwm_init(3, BL_PWM_FREQ_DEFAULT);
+#endif
+
+#ifdef BOARD_PWM_CH4_PIN
```
[the remark this answers is not in the document]
Comment (liangzhanggb) [reply arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c]: I have fixed it, Thank you very much

File: arch/risc-v/src/bl602/Make.defs [hunk 4]
```diff
@@ -58,7 +54,12 @@ endif
 ifeq ($(CONFIG_ONESHOT),y)
 CHIP_CSRCS  += bl602_oneshot_lowerhalf.c
 endif
-
+ifeq ($(CONFIG_WDOG),y)
```
[the remark this answers is not in the document]
Comment (liangzhanggb) [reply arch/risc-v/src/bl602/Make.defs]: I have fixed it, Thank you very much

**Comment by btashton** [comment]
@liangzhanggb One more minor fix that I think will also get the build to pass.  I verified your latest changes and they all look good.  Thanks.
```c
diff --git a/arch/risc-v/src/bl602/Kconfig b/arch/risc-v/src/bl602/Kconfig
index 9542d0be9c..0c819f95c8 100644
--- a/arch/risc-v/src/bl602/Kconfig
+++ b/arch/risc-v/src/bl602/Kconfig
@@ -13,6 +13,7 @@ config BL602_HAVE_UART0
        select ARCH_HAVE_UART0
        select UART0_SERIALDRIVER
        select ARCH_HAVE_SERIAL_TERMIOS
+       select ARCH_HAVE_PWM_MULTICHAN
 
 config BL602_UART0
        bool
@@ -37,6 +38,5 @@ config BL602_TIMER1
 
 config BL602_PWM0
        bool "PWM0"
-       select ARCH_HAVE_PWM_MULTICHAN
 
 endmenu
diff --git a/boards/risc-v/bl602/bl602evb/configs/timer/defconfig b/boards/risc-v/bl602/bl602evb/configs/timer/defconfig
index 63ef88f96a..94d63251ff 100644
--- a/boards/risc-v/bl602/bl602evb/configs/timer/defconfig
+++ b/boards/risc-v/bl602/bl602evb/configs/timer/defconfig
@@ -55,6 +55,7 @@ CONFIG_ONESHOT=y
 CONFIG_PREALLOC_TIMERS=0
 CONFIG_PTHREAD_STACK_DEFAULT=8192
 CONFIG_PWM=y
+CONFIG_PWM_MULTICHAN=y
 CONFIG_PWM_NCHANNELS=2
 CONFIG_RAM_SIZE=134217728
 CONFIG_RAM_START=0xc0800000
```

**Comment by liangzhanggb** [comment]
> @liangzhanggb One more minor fix that I think will also get the build to pass. I verified your latest changes and they all look good. Thanks.
> 
> ```c
> diff --git a/arch/risc-v/src/bl602/Kconfig b/arch/risc-v/src/bl602/Kconfig
> index 9542d0be9c..0c819f95c8 100644
> --- a/arch/risc-v/src/bl602/Kconfig
> +++ b/arch/risc-v/src/bl602/Kconfig
> @@ -13,6 +13,7 @@ config BL602_HAVE_UART0
>         select ARCH_HAVE_UART0
>         select UART0_SERIALDRIVER
>         select ARCH_HAVE_SERIAL_TERMIOS
> +       select ARCH_HAVE_PWM_MULTICHAN
>  
>  config BL602_UART0
>         bool
> @@ -37,6 +38,5 @@ config BL602_TIMER1
>  
>  config BL602_PWM0
>         bool "PWM0"
> -       select ARCH_HAVE_PWM_MULTICHAN
>  
>  endmenu
> diff --git a/boards/risc-v/bl602/bl602evb/configs/timer/defconfig b/boards/risc-v/bl602/bl602evb/configs/timer/defconfig
> index 63ef88f96a..94d63251ff 100644
> --- a/boards/risc-v/bl602/bl602evb/configs/timer/defconfig
> +++ b/boards/risc-v/bl602/bl602evb/configs/timer/defconfig
> @@ -55,6 +55,7 @@ CONFIG_ONESHOT=y
>  CONFIG_PREALLOC_TIMERS=0
>  CONFIG_PTHREAD_STACK_DEFAULT=8192
>  CONFIG_PWM=y
> +CONFIG_PWM_MULTICHAN=y
>  CONFIG_PWM_NCHANNELS=2
>  CONFIG_RAM_SIZE=134217728
>  CONFIG_RAM_START=0xc0800000
> ```

I have fixed it, Thank you very much

**Comment by btashton** [comment]
LGTM I see a couple other unrelated things for this port in the Kconfigs that I can clean in another PR.

File: arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+  int                     ret  = OK;
+  int                     i;
+
+  UNUSED(i);
+
+#ifdef CONFIG_PWM_NCHANNELS
+  for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+    {
+      bl602_pwm_freq(priv, i, info->frequency);
+      bl602_pwm_duty(priv, i, info->channels[i].duty);
```
Comment (normanr) [inline arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c:376]: Should the second argument have been `info->channels[i].channel`? Otherwise there's no way to change the order of channels.

Also PR #4227 added an option to break these loops when `channel == -1`, it looks like it was missed here.

File: arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+#define BL_PWM_XTAL_CLK           (40000000)
+#define BL_PWM_BUS_BCLK           (80000000)
+
+#ifdef CONFIG_PWM_NCHANNELS
+#  if (CONFIG_PWM_NCHANNELS > 5)
+#    error "The maximum number of BL602 PWM channels is 5!"
+#  endif
+#endif
+
+#ifdef CONFIG_BL602_PWM0_BUS_BCLK_SRC
```
Comment (normanr) [inline arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c:59]: Is `CONFIG_BL602_PWM0_BUS_BCLK_SRC` configured anywhere? I expected to see it in a Kconfig, but can't find it.

Comment (xiaoxiang781216) [reply arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c:59]: @liangzhanggb please take a look.

File: arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c (hunk 5 above)
Comment (normanr) [reply arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c:376]: PR https://github.com/apache/incubator-nuttx/pull/5091 to add the early exit from PR #4227.

Comment (normanr) [reply arch/risc-v/src/bl602/bl602_pwm_lowerhalf.c:376]: and PR https://github.com/apache/incubator-nuttx/pull/5093 to use `chan` instead of `i`.
